File: du-backend/src/main/java/com/dwin/du/update/EntityUpdateService.java
package com.dwin.du.update;
import com.dwin.du.api.entity.*;
import com.dwin.du.api.repository.*;
import com.dwin.du.validation.ValidationService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
@RequiredArgsConstructor
public class EntityUpdateService {
    private final PersonRepository personRepository;
    private final ReceiptRepository receiptRepository;
    private final ProductRepository productRepository;
    private final PersonProductRepository personProductRepository;
    private final LoanRepository loanRepository;
    private final ValidationService validator;

    // Person
    public void updatePerson(String username, Boolean updateBalance) {
        User user = validator.validateUser(username);
        List<Person> persons = personRepository.findByUser(user);

        for (Person person : persons) {
            List<PersonProduct> personProducts = personProductRepository.findByPerson(person);

            long receiptsCount = personProducts.stream()
                    .map(personProduct -> personProduct.getProduct())
                    .filter(product -> product.getReceipt() != null)
                    .map(product -> product.getReceipt().getId())
                    .distinct()
                    .count();

            person.setReceiptsCount((int) receiptsCount);

            person.setProductsCount(personProducts.size());

            double totalAmount = personProducts.stream()
                    .mapToDouble(PersonProduct::getPartOfPrice)
                    .sum();
            person.setTotalAmount(totalAmount);

            double unpaidAmount = personProducts.stream()
                    .filter(personProduct -> !personProduct.isSettled())
                    .mapToDouble(PersonProduct::getPartOfPrice)
                    .sum();
            person.setUnpaidAmount(unpaidAmount);

            if(updateBalance){
                double balance = calculateBalance(person);
                person.setLoanBalance(balance);
            }
        }

        personRepository.saveAll(persons);
    }

    private double calculateBalance(Person person) {
        List<Loan> loans = loanRepository.findByPerson(person);
        double balance = 0.0;

        for (Loan loan : loans) {
            if(!loan.isSettled()){
                if (loan.isLent())
                    balance += loan.getAmount();
                else
                    balance -= loan.getAmount();
            }
        }

        return balance;
    }

    // Receipt
    public void updateTotalPriceReceipt(Receipt receipt) {
        List<Product> products = productRepository.findByReceipt(receipt);
        double totalPrice = 0.00;
        for(var item : products)
            totalPrice += item.getPrice();

        receipt.setTotalPrice(totalPrice);
        receiptRepository.save(receipt);
    }

    public boolean areAllProductsSettled(Receipt receipt) {
        List<Product> products = productRepository.findByReceipt(receipt);
        return products.stream().allMatch(Product::isSettled);
    }

    // Product
    public void updateCompensationPrice(Product product) {
        List<PersonProduct> personProducts = personProductRepository.findByProduct(product);
        double compensationPrice = calculateCompensationPrice(personProducts, product.getPrice());
        product.setCompensationPrice(compensationPrice);
        productRepository.save(product);
    }
    public double calculateCompensationPrice(List<PersonProduct> personProductDtos, double price){
        double compensationPrice = 0;
        for(var item : personProductDtos){
            compensationPrice += item.getPartOfPrice();
        }
        compensationPrice = price - compensationPrice;
        return compensationPrice;
    }

    public boolean areAllPersonProductsSettled(Product product) {
        List<PersonProduct> personProducts = personProductRepository.findByProduct(product);
        return personProducts.stream().allMatch(PersonProduct::isSettled);
    }

    // Person product
    public void updatePartPricesPersonProduct(Product product) {
        List<PersonProduct> personProducts = personProductRepository.findByProduct(product);
        for(var item : personProducts){
            double partPrice = calculatePartPrice(item.getQuantity(), product.getMaxQuantity(), product.getPrice());
            item.setPartOfPrice(partPrice);
            personProductRepository.save(item);
        }
    }

    public double calculatePartPrice(int quantity, int maxQuantity, double price) {
        double partPrice = ((double) quantity / maxQuantity) * price;
        return partPrice;
    }
}

File: divvyup/DivvyUp.Web/Update/EntityUpdateService.cs
using DivvyUp.Web.Data;
using DivvyUp_Shared.Models;
using Microsoft.EntityFrameworkCore;

namespace DivvyUp.Web.Update
{
    public class EntityUpdateService
    {
        private readonly DivvyUpDBContext _dbContext;

        public EntityUpdateService(DivvyUpDBContext dbContext)
        {
            _dbContext = dbContext;
        }

        public async Task UpdatePerson(User user, bool updateBalance)
        {
            var persons = await _dbContext.Persons
                .Where(p => p.UserId == user.Id)
                .Include(p => p.PersonProducts)
                .ThenInclude(pp => pp.Product)
                .ToListAsync();

            foreach (var person in persons)
            {
                var personProducts = person.PersonProducts;
                var receipts = personProducts
                    .Select(pp => pp.Product.ReceiptId)
                    .Distinct()
                    .ToList();

                person.ReceiptsCount = receipts.Count;
                person.ProductsCount = personProducts.Count;
                person.TotalAmount = personProducts.Sum(pp => pp.PartOfPrice + (pp.Compensation ? pp.Product.CompensationPrice : 0));
                person.UnpaidAmount = personProducts.Where(pp => !pp.Settled).Sum(pp => pp.PartOfPrice + (pp.Compensation ? pp.Product.CompensationPrice : 0));

                if (updateBalance)
                {
                    person.LoanBalance = await CalculateBalance(person);
                }
            }

            _dbContext.Persons.UpdateRange(persons);
            await _dbContext.SaveChangesAsync();
        }

        private async Task<decimal> CalculateBalance(Person person)
        {
            return await _dbContext.Loans
                .Where(l => l.PersonId == person.Id && !l.Settled)
                .SumAsync(l => l.Lent ? l.Amount : -l.Amount);
        }


        public async Task UpdateTotalPriceReceipt(Receipt receipt)
        {
            receipt.TotalPrice = await _dbContext.Products
                .Where(p => p.ReceiptId == receipt.Id)
                .SumAsync(p => p.Price);

            _dbContext.Receipts.Update(receipt);
            await _dbContext.SaveChangesAsync();
        }

        public async Task<bool> AreAllProductsSettled(Receipt receipt)
        {
            var products = await _dbContext.Products
                .Where(p => p.Receipt.Id == receipt.Id)
                .ToListAsync();
            return products.All(p => p.Settled);
        }

        public async Task UpdateProductDetails(Product product)
        {
            var personProducts = await _dbContext.PersonProducts
                .Where(pp => pp.ProductId == product.Id)
                .ToListAsync();

            var compensationPrice = product.Price - personProducts.Sum(pp => pp.PartOfPrice);
            product.CompensationPrice = compensationPrice;

            var availableQuantity = product.MaxQuantity - personProducts.Sum(pp => pp.Quantity);
            product.AvailableQuantity = availableQuantity;

            _dbContext.Products.Update(product);
            await _dbContext.SaveChangesAsync();
        }


        public async Task<bool> AreAllPersonProductsSettled(Product product)
        {
            var personProducts = await _dbContext.PersonProducts
                .Where(pp => pp.Product.Id == product.Id)
                .ToListAsync();
            return personProducts.All(pp => pp.Settled);
        }

        public async Task UpdatePartPricesPersonProduct(Product product)
        {
            var personProducts = await _dbContext.PersonProducts
                .Where(pp => pp.ProductId == product.Id)
                .ToListAsync();

            bool isUpdated = false;

            foreach (var personProduct in personProducts)
            {
                var newPartOfPrice = await CalculatePartPrice(personProduct.Quantity, product.MaxQuantity, product.Price);
                if (personProduct.PartOfPrice != newPartOfPrice)
                {
                    personProduct.PartOfPrice = newPartOfPrice;
                    isUpdated = true;
                }
            }

            if (isUpdated)
            {
                _dbContext.PersonProducts.UpdateRange(personProducts);
                await _dbContext.SaveChangesAsync();
            }
        }

        private Task<decimal> CalculatePartPrice(int quantity, int maxQuantity, decimal price)
        {
            return Task.FromResult(((decimal)quantity / maxQuantity) * price);
        }
    }
}
